Assignment 7

task 1

Given the following data from an e-commerce app: out of 500 users, 120 bought phone accessories, 80 bought a phone, and 50 bought both, calculate the probability that a user bought a phone given they already bought accessories (P(Phone|Accessories)).

In [1]:
import numpy as np

total_users = 500
accessories_buyers = 120
phone_buyers = 80
both_buyers = 50

p_accessories = accessories_buyers / total_users
p_phone_and_accessories = both_buyers / total_users

p_phone_given_accessories = np.divide(p_phone_and_accessories, p_accessories)

print(f"P(Accessories) = {accessories_buyers}/{total_users} = {p_accessories}")
print(f"P(Phone and Accessories) = {both_buyers}/{total_users} = {p_phone_and_accessories}")
print(f"P(Phone | Accessories) = {p_phone_and_accessories} / {p_accessories} = {round(p_phone_given_accessories, 4)}")


P(Accessories) = 120/500 = 0.24
P(Phone and Accessories) = 50/500 = 0.1
P(Phone | Accessories) = 0.1 / 0.24 = 0.4167


**Explanation:** Using the conditional probability formula P(A|B) = P(A and B) / P(B), we treat "Phone" as event A and "Accessories" as event B. P(Accessories) = 120/500 = 0.24 and P(Phone and Accessories) = 50/500 = 0.10. Dividing gives P(Phone|Accessories) = 0.10 / 0.24 ≈ 0.4167, or about 41.7%. This is much higher than the overall rate of buying a phone (80/500 = 16%), showing that customers who already bought accessories are significantly more likely to also buy a phone — useful for targeted upselling.

task 2

Imagine you are analyzing Flipkart's user data: 30% of users buy headphones, and among those, 40% later buy a smartphone. Calculate the conditional probability that a randomly selected user buys a smartphone, given they have already bought headphones.

**Hint:** Use the formula for conditional probability: P(A|B) = P(A and B) / P(B).

In [2]:
import numpy as np

p_headphones = 0.30
p_smartphone_given_headphones = 0.40  # given directly in the problem

# Bonus: joint probability P(A and B) = P(A|B) * P(B)
p_smartphone_and_headphones = np.multiply(p_smartphone_given_headphones, p_headphones)

print(f"P(Headphones) = {p_headphones}")
print(f"P(Smartphone | Headphones) = {p_smartphone_given_headphones}")
print(f"P(Smartphone and Headphones) = {p_smartphone_given_headphones} * {p_headphones} = {p_smartphone_and_headphones}")


P(Headphones) = 0.3
P(Smartphone | Headphones) = 0.4
P(Smartphone and Headphones) = 0.4 * 0.3 = 0.12


**Explanation:** The problem already states the conditional probability directly: "among those [headphone buyers], 40% later buy a smartphone" means P(Smartphone | Headphones) = 0.40. This is the direct answer to what's asked. As a bonus, we compute the joint probability that a random user buys both: P(Smartphone and Headphones) = P(Smartphone|Headphones) x P(Headphones) = 0.40 x 0.30 = 0.12, meaning 12% of all users buy both headphones and a smartphone (in that order).

task 3

Write a Python function bayes_churn_probability(prior_churn, prob_email_given_churn, prob_email) that uses Bayes' Theorem to compute the probability a user will churn given they received a 'we miss you' email. Test your function with prior_churn=0.15, prob_email_given_churn=0.6, prob_email=0.25.

In [3]:
import numpy as np

def bayes_churn_probability(prior_churn, prob_email_given_churn, prob_email):
    """
    Applies Bayes' Theorem:
    P(Churn | Email) = [P(Email | Churn) * P(Churn)] / P(Email)
    """
    numerator = np.multiply(prob_email_given_churn, prior_churn)
    return np.divide(numerator, prob_email)

# Test the function with the given values
result = bayes_churn_probability(
    prior_churn=0.15,
    prob_email_given_churn=0.6,
    prob_email=0.25
)

print(f"P(Churn | Email) = {round(float(result), 4)}")


P(Churn | Email) = 0.36


**Explanation:** Bayes' Theorem lets us update our belief about churn once we observe a related piece of evidence (receiving the "we miss you" email). Plugging in the numbers: P(Churn|Email) = (0.6 x 0.15) / 0.25 = 0.09 / 0.25 = 0.36. So a user who received the email has a 36% chance of churning, compared to the baseline 15% churn rate. This tells the marketing team that receiving this email is strongly associated with higher churn risk — likely because it is being sent precisely to users already showing disengagement signals.

task 4

A digital wallet app detects fraud based on spending patterns. If 2% of all transactions are fraudulent, and the app's alert system correctly flags 90% of fraudulent transactions but also falsely flags 5% of normal transactions, use Bayes' Theorem to calculate the probability that a transaction is actually fraudulent when flagged by the alert system.

**Hint:** Define events: F = fraud, A = alert. Use Bayes' formula: P(F|A) = [P(A|F) * P(F)] / P(A).

In [4]:
import numpy as np

p_fraud = 0.02                 # P(F): overall fraud rate
p_not_fraud = 1 - p_fraud      # P(not F)

p_alert_given_fraud = 0.90     # P(A|F): true positive rate
p_alert_given_not_fraud = 0.05 # P(A|not F): false positive rate

# Step 1: total probability of an alert (law of total probability)
p_alert = np.add(p_alert_given_fraud * p_fraud, p_alert_given_not_fraud * p_not_fraud)

# Step 2: apply Bayes' Theorem
p_fraud_given_alert = np.divide(p_alert_given_fraud * p_fraud, p_alert)

print(f"P(Fraud) = {p_fraud}")
print(f"P(Alert) = {round(float(p_alert), 4)}")
print(f"P(Fraud | Alert) = {round(float(p_fraud_given_alert), 4)}")


P(Fraud) = 0.02
P(Alert) = 0.067
P(Fraud | Alert) = 0.2687


**Explanation:** First we find P(Alert) using the law of total probability: P(A) = P(A|F)xP(F) + P(A|not F)xP(not F) = (0.90 x 0.02) + (0.05 x 0.98) = 0.018 + 0.049 = 0.067. Then Bayes' Theorem gives P(Fraud|Alert) = 0.018 / 0.067 ≈ 0.2687, or about 26.9%. This is a classic illustration of the base-rate fallacy: even though the alert system catches 90% of actual fraud, only about 27% of *flagged* transactions turn out to be fraudulent, because genuine fraud is so rare (2%) that the much larger pool of normal transactions generates a lot of false alarms.

task 5

Use ChatGPT or another AI tool to generate a real-world scenario where conditional probability is used in a food delivery app (like Zomato or Swiggy), and explain how Bayes' Theorem could help improve their recommendation system. Paste the AI's output and add one sentence of your own analysis.

In [5]:
ai_generated_scenario = (
    "Swiggy wants to predict whether a user will order dessert given that they've already "
    "ordered from a South Indian restaurant. Historical data shows that 25% of all orders "
    "include a South Indian restaurant, and among those, 35% also include a dessert add-on, "
    "compared to only 12% dessert attachment on orders overall. Using P(Dessert | South Indian) "
    "= 0.35, the app can prioritize dessert recommendations specifically to users mid-order from "
    "South Indian restaurants, rather than showing generic dessert prompts to everyone. Bayes' "
    "Theorem becomes useful here when the app wants to reverse the question - for example, given "
    "that a user added a dessert, what's the probability they ordered from a South Indian "
    "restaurant first? - which helps with restaurant-side promotional targeting."
)

my_analysis = (
    "This shows how Bayes' Theorem helps flip an easy-to-measure probability (cuisine leads to "
    "add-on) into a harder-to-measure one (add-on tells us about likely cuisine), which is very "
    "useful for targeted recommendations and restaurant-side promotions."
)

print("AI-generated scenario:\n")
print(ai_generated_scenario)
print("\nMy own analysis:\n")
print(my_analysis)


AI-generated scenario:

Swiggy wants to predict whether a user will order dessert given that they've already ordered from a South Indian restaurant. Historical data shows that 25% of all orders include a South Indian restaurant, and among those, 35% also include a dessert add-on, compared to only 12% dessert attachment on orders overall. Using P(Dessert | South Indian) = 0.35, the app can prioritize dessert recommendations specifically to users mid-order from South Indian restaurants, rather than showing generic dessert prompts to everyone. Bayes' Theorem becomes useful here when the app wants to reverse the question - for example, given that a user added a dessert, what's the probability they ordered from a South Indian restaurant first? - which helps with restaurant-side promotional targeting.

My own analysis:

This shows how Bayes' Theorem helps flip an easy-to-measure probability (cuisine leads to add-on) into a harder-to-measure one (add-on tells us about likely cuisine), which i

**Explanation:** This task is conceptual rather than numerical, so the scenario and analysis are simply stored as text strings and printed. The AI-generated example shows P(Dessert | South Indian) = 0.35 being used directly for recommendations, while the one-line analysis explains why Bayes' Theorem's reverse-direction reasoning is valuable for recommendation systems.